File: ts/src/rlm/session.ts
/**
 * RLM Session — drives the multi-turn REPL conversation loop for one agent role.
 * Mirrors Python autocontext.harness.repl.session.RlmSession.
 */

import type { ReplWorker, LlmComplete, ExecutionRecord } from "./types.js";

export interface RlmSessionOpts {
  complete: LlmComplete;
  worker: ReplWorker;
  role: string;
  model: string;
  systemPrompt: string;
  initialUserMessage?: string;
  maxTurns?: number;
  maxTokensPerTurn?: number;
  temperature?: number;
  onTurn?: (current: number, total: number, ready: boolean) => void;
}

export interface RlmResult {
  content: string;
  executionHistory: ExecutionRecord[];
  turnsUsed: number;
}

/** Extract code from the first code block delimited by code tags. */
export function extractCode(text: string): string | null {
  const match = text.match(/<code>([\s\S]*?)<\/code>/);
  return match ? match[1].trim() : null;
}

/**
 * Drives the multi-turn REPL conversation loop for one agent role.
 *
 * Flow per turn:
 * 1. Send conversation history to LLM
 * 2. Extract code block from response
 * 3. Run code via worker
 * 4. Build feedback from stdout/error
 * 5. Check answer["ready"] flag — if true, exit loop
 * 6. Repeat until maxTurns
 */
export class RlmSession {
  private readonly complete: LlmComplete;
  private readonly worker: ReplWorker;
  private readonly role: string;
  private readonly model: string;
  private readonly systemPrompt: string;
  private readonly initialUserMessage: string;
  private readonly maxTurns: number;
  private readonly maxTokensPerTurn: number;
  private readonly temperature: number;
  private readonly onTurn?: RlmSessionOpts["onTurn"];

  constructor(opts: RlmSessionOpts) {
    this.complete = opts.complete;
    this.worker = opts.worker;
    this.role = opts.role;
    this.model = opts.model;
    this.systemPrompt = opts.systemPrompt;
    this.initialUserMessage = opts.initialUserMessage ?? "Begin exploring the data.";
    this.maxTurns = opts.maxTurns ?? 15;
    this.maxTokensPerTurn = opts.maxTokensPerTurn ?? 2048;
    this.temperature = opts.temperature ?? 0.2;
    this.onTurn = opts.onTurn;
  }

  /** Run the full REPL loop and return an RlmResult. */
  async run(): Promise<RlmResult> {
    const messages: Array<{ role: string; content: string }> = [
      { role: "user", content: this.initialUserMessage },
    ];
    const executionHistory: ExecutionRecord[] = [];
    let finalContent = "";
    let answeredReady = false;

    for (let turn = 1; turn <= this.maxTurns; turn++) {
      const response = await this.complete(messages, {
        model: this.model,
        maxTokens: this.maxTokensPerTurn,
        temperature: this.temperature,
        systemPrompt: this.systemPrompt,
      });

      const assistantText = response.text;
      messages.push({ role: "assistant", content: assistantText });

      const code = extractCode(assistantText);
      if (code === null) {
        messages.push({
          role: "user",
          content:
            'Please write code inside <code> tags to continue your analysis, or set answer["ready"] = True to finalize.',
        });
        this.onTurn?.(turn, this.maxTurns, false);
        continue;
      }

      const result = await this.worker.runCode({ code });
      const answerReady = result.answer?.["ready"] === true;

      executionHistory.push({
        turn,
        code,
        stdout: result.stdout,
        error: result.error,
        answerReady,
      });

      const parts: string[] = [];
      if (result.stdout) {
        parts.push(`Output:\n${result.stdout}`);
      }
      if (result.error) {
        parts.push(`Error:\n${result.error}`);
      }
      if (parts.length === 0) {
        parts.push("(no output)");
      }
      const feedback = parts.join("\n");

      this.onTurn?.(turn, this.maxTurns, answerReady);

      if (answerReady) {
        finalContent = String(result.answer?.["content"] ?? "");
        answeredReady = true;
        break;
      }

      messages.push({ role: "user", content: feedback });
    }

    if (!answeredReady && this.worker.namespace?.["answer"]) {
      const ans = this.worker.namespace["answer"] as Record<string, unknown>;
      finalContent = String(ans?.["content"] ?? "");
    }

    return {
      content: finalContent,
      executionHistory,
      turnsUsed: executionHistory.length,
    };
  }
}

File: autocontext/src/autocontext/harness/repl/session.py
"""Domain-agnostic REPL session for multi-turn LLM exploration."""

from __future__ import annotations

import logging
import re
import time
import uuid
from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor
from typing import Any

from autocontext.harness.core.llm_client import LanguageModelClient
from autocontext.harness.core.types import RoleExecution, RoleUsage
from autocontext.harness.repl.types import ExecutionRecord, ReplCommand, ReplWorkerProtocol

logger = logging.getLogger(__name__)

_CODE_PATTERN = re.compile(r"<code>(.*?)</code>", re.DOTALL)


def make_llm_batch(
    client: LanguageModelClient,
    model: str,
    max_tokens: int = 1024,
    temperature: float = 0.1,
    max_workers: int = 4,
) -> Callable[[list[str]], list[str]]:
    """Create an ``llm_batch()`` callable for injection into the REPL namespace."""

    def llm_batch(prompts: list[str]) -> list[str]:
        if not prompts:
            return []
        workers = min(len(prompts), max_workers)
        with ThreadPoolExecutor(max_workers=workers) as pool:
            futures = [
                pool.submit(
                    client.generate,
                    model=model,
                    prompt=p,
                    max_tokens=max_tokens,
                    temperature=temperature,
                )
                for p in prompts
            ]
            results: list[str] = []
            for f in futures:
                try:
                    results.append(f.result().text)
                except Exception as exc:  # noqa: BLE001
                    logger.debug("harness.repl.session: caught Exception", exc_info=True)
                    results.append(f"[llm_batch error: {exc}]")
            return results

    return llm_batch


class RlmSession:
    """Drives the multi-turn REPL conversation loop for one agent role."""

    def __init__(
        self,
        client: LanguageModelClient,
        worker: ReplWorkerProtocol,
        role: str,
        model: str,
        system_prompt: str,
        initial_user_message: str = "Begin exploring the data.",
        max_turns: int = 15,
        max_tokens_per_turn: int = 2048,
        temperature: float = 0.2,
        on_turn: Callable[[int, int, bool], None] | None = None,
    ) -> None:
        self._client = client
        self._worker = worker
        self._role = role
        self._model = model
        self._system = system_prompt
        self._initial_msg = initial_user_message
        self._max_turns = max_turns
        self._max_tokens = max_tokens_per_turn
        self._temperature = temperature
        self._on_turn = on_turn
        self.execution_history: list[ExecutionRecord] = []

    def run(self) -> RoleExecution:
        """Execute the full REPL loop and return a RoleExecution."""
        started = time.perf_counter()
        messages: list[dict[str, str]] = [{"role": "user", "content": self._initial_msg}]
        total_input = 0
        total_output = 0
        status = "completed"

        def _get_history() -> list[dict[str, Any]]:
            return [
                {
                    "turn": r.turn,
                    "code_preview": r.code[:200],
                    "stdout_preview": r.stdout[:200],
                    "error": r.error,
                }
                for r in self.execution_history
            ]

        self._worker.namespace["get_history"] = _get_history

        for turn in range(1, self._max_turns + 1):
            response = self._client.generate_multiturn(
                model=self._model,
                system=self._system,
                messages=messages,
                max_tokens=self._max_tokens,
                temperature=self._temperature,
            )
            total_input += response.usage.input_tokens
            total_output += response.usage.output_tokens

            assistant_text = response.text
            messages.append({"role": "assistant", "content": assistant_text})

            code_match = _CODE_PATTERN.search(assistant_text)
            if code_match:
                code = code_match.group(1).strip()
                result = self._worker.run_code(ReplCommand(code))

                self.execution_history.append(ExecutionRecord(
                    turn=turn,
                    code=code,
                    stdout=result.stdout,
                    error=result.error,
                    answer_ready=result.answer.get("ready", False),
                ))

                # Build user feedback message
                parts: list[str] = []
                if result.stdout:
                    parts.append(f"[stdout]\n{result.stdout}")
                if result.error:
                    parts.append(f"[error]\n{result.error}")
                if not parts:
                    parts.append("[no output]")

                feedback = "\n\n".join(parts)
                messages.append({"role": "user", "content": feedback})

                if self._on_turn:
                    self._on_turn(turn, self._max_turns, result.answer.get("ready", False))

                if result.answer.get("ready"):
                    logger.debug("RLM %s finished on turn %d", self._role, turn)
                    break
            else:
                # Model didn't emit code — nudge it
                messages.append({
                    "role": "user",
                    "content": "Please write code inside <code> tags to continue your analysis, "
                    'or set answer["ready"] = True to finalize.',
                })
        else:
            status = "truncated"
            logger.warning("RLM %s hit max_turns=%d without finalizing", self._role, self._max_turns)

        answer = self._worker.namespace.get("answer", {"content": "", "ready": False})
        content = answer.get("content", "")
        elapsed_ms = int((time.perf_counter() - started) * 1000)

        return RoleExecution(
            role=self._role,
            content=content,
            usage=RoleUsage(
                input_tokens=total_input,
                output_tokens=total_output,
                latency_ms=elapsed_ms,
                model=self._model,
            ),
            subagent_id=uuid.uuid4().hex[:10],
            status=status,
        )
